### PR TITLE
Bug 1167344 - added support for filtering add-on's based on the app.

### PR DESCRIPTION
**Pull request by yzen**
Change-Id: I035d87c4cbb1802d07f92f5ca56453e207bf5903

### CODE REVIEW

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d271d3f7966f63197bcff5470a5e11adb94b9c72

File: apps/settings/js/modules/addon_manager.js [hunk 1]
```diff
@@ -205,71 +205,114 @@ define(function(require) {
...
+     * and also derive the 'reboot' and 'restart' return values of
+     * disableAddon() from it.
+     *
+     * @access private
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     * @param {DOMApplication} app
+     * @returns {Promise}
+     */
+    _getCustomizedApp: function(addon, app) {
```
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: The function name needs to be updated as the behavior changes. The function now act more like "get the customizations of an addon that are being applied on an app".

File: apps/settings/js/modules/addon_manager.js [hunk 2]
```diff
@@ -205,71 +205,114 @@ define(function(require) {
...
-          if (appliedCustomizations.length > 0) {
-            customizedApps.push({
-              app: app,
-              customizations: appliedCustomizations
-            });
-          }
+      // If any customizations were applied to this app, return the list of
+      // appliedCustomizations along with the app
+      if (appliedCustomizations.length > 0) {
+        return Promise.resolve({
```
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: I think we can resolve with only `applideCustomizations` as `app` and `addon` are passed as parameters.

File: apps/settings/js/modules/addon_manager.js [hunk 3]
```diff
@@ -63,6 +63,10 @@ define(function(require) {
       if (this._isAddon(app)) {
         if (type === 'install' && !this._alreadyExists(app)) {
           this._addons.push(App(app));
+          // If there's an app based filter, apply it.
+          if (this._filter) {
+            this._applyFilter();
```
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: I would suggest to do the filtering outside of the addon manager, which give us more flexibility on having different types of filters.

File: apps/settings/js/modules/addon_manager.js (hunk 1 above)
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: This function can be considered as a utility function exposed to others for checking if there are any customization acting on an app. Then we are able to use it in our filter.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=41d99faf79c346abdfa1b6a8bd52cdc8a9d3aed8

File: apps/settings/js/modules/addon_manager.js [hunk 4]
```diff
@@ -205,71 +201,80 @@ define(function(require) {
...
+     * values of disableAddon() from it and also use this function to apply
+     * addon filters.
+     *
+     * @access private
+     * @memberOf AddonManager.prototype
+     * @param {DOMApplication} addon
+     * @param {DOMApplication} app
+     * @returns {Promise}
+     */
+    _getAddonAppCustomizations: function(addon, app) {
```
Comment (crh0716) [inline apps/settings/js/modules/addon_manager.js]: This is a synchronous function, it is not required to return a promise.

File: apps/settings/js/panels/addons/addons_list.js [hunk 5]
```diff
@@ -9,16 +9,22 @@ define(function(require) {
   var SettingsService = require('modules/settings_service');
   var ListView = require('modules/mvvm/list_view');
   var AddonItemTemplateFactory = require('panels/addons/addons_template');
+  var ObservableArray = require('modules/mvvm/observable_array');
 
   function AddonsList(root, manager) {
     this._enabled = false;
     this._listRoot = root;
     this._addonManager = manager;
+    this._addons = ObservableArray(this._addonManager.addons.array);
```
Comment (crh0716) [inline apps/settings/js/panels/addons/addons_list.js]: Let's initialize it with an empty array and call to `_update` for the real initialization.

**Comment by lightsofapollo** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=d1709e1363ddd630b671d7ec30f8252e6ab1ec6f

**Pull request merged**